src/bin/cql-stress-cassandra-stress/java_generate/mod.rs

```diff
@@ -1,3 +1,56 @@
 pub mod distribution;
 pub mod faster_random;
 pub mod values;
+
+/// A structure wrapping [java_random::Random].
+/// In addition to methods exposed by [java_random::Random],
+/// it exposes [next_gaussian] method as well.
+struct Random {
```

Comment:
> Notice: Java's implementation of this function makes use of `StrictMath`,
> which may give different results for square root and natural log functions for floats.
> However, the results shouldn't differ when we cast the them to integers.

I wouldn't be so sure. If the result of `next_gaussian()` is close to an integral value and both implementations return slightly different results - so that one returns a value below the integer and the other one above it - then you will get different results when casting to an integer.

How important it is to have exactly the same results as the Java's `Random`?

Reply:
Yeah... I am aware of the fact that there is a non-zero probability that the samples from Rust and Java may differ (even though it's very small).

The only scenario (I can think of) where cql-stress might be inconsistent with c-s is:
```
cql-stress write n=10 -col 'size=GAUSSIAN(1..100)'
cassandra-stress read n=10 -col 'size=GAUSSIAN(1..100)'
```
Assume that within 10 [number of ops] * 5 [default number of blob columns in row] = 50 samples from the GAUSSIAN(1..100) distribution, there is some `i` such that `cql_stress_samples[i] != c_s_samples[i]`.
Then we get a cassandra-stress read error (data was not successfully validated).

Notice that `-col size=` (the distribution from which we sample the size of the blob cell) is the only distribution that can be set by the user, and in addition, samples from this distribution are deterministic (since we [seed](https://github.com/scylladb/cql-stress/blob/master/src/bin/cql-stress-cassandra-stress/java_generate/values/blob.rs#L23) the distribution before each sample).

The other distribution that we use in a deterministic way (set seed before each sample), is an `identity_distribution`. This distribution, however, has a default value of ['UNIFORM(1..100B)'](https://github.com/scylladb/cql-stress/blob/master/src/bin/cql-stress-cassandra-stress/java_generate/values/mod.rs#L56) and cannot be set by the user via CLI.

Now the question is, whether we want to import an additional crate with `StrictMath` (e.g. https://crates.io/crates/rust-strictmath), just for the sake of Gaussian distribution behaving in the same way as Java's implementation does.

NOTICE: this may also be the problem in the future once we decide to seed `pk_seed` distribution with operation's id. This distribution is set via `-pop dist=` parameter, and is used to sample the seeds for `partition_key` generator for each operation.

Reply:
> Yeah... I am aware of the fact that there is a non-zero probability that the samples from Rust and Java may differ (even though it's very small).
> 
> The only scenario (I can think of) where cql-stress might be inconsistent with c-s is:
> 
> ```
> cql-stress write n=10 -col 'size=GAUSSIAN(1..100)'
> cassandra-stress read n=10 -col 'size=GAUSSIAN(1..100)'
> ```
> 
> Assume that within 10 [number of ops] * 5 [default number of blob columns in row] = 50 samples from the GAUSSIAN(1..100) distribution, there is some `i` such that `cql_stress_samples[i] != c_s_samples[i]`. Then we get a cassandra-stress read error (data was not successfully validated).

Isn't `[number of ops]` going to be much higher in practice, i.e. in SCT tests? If so, then the probability rises. If we are going to write with our implementation of c-s and then validate with the original c-s, then this is prone to flakiness.

> 
> Notice that `-col size=` (the distribution from which we sample the size of the blob cell) is the only distribution that can be set by the user, and in addition, samples from this distribution are deterministic (since we [seed](https://github.com/scylladb/cql-stress/blob/master/src/bin/cql-stress-cassandra-stress/java_generate/values/blob.rs#L23) the distribution before each sample).
> 
> The other distribution that we use in a deterministic way (set seed before each sample), is an `identity_distribution`. This distribution, however, has a default value of ['UNIFORM(1..100B)'](https://github.com/scylladb/cql-stress/blob/master/src/bin/cql-stress-cassandra-stress/java_generate/values/mod.rs#L56) and cannot be set by the user via CLI.
> 
> Now the question is, whether we want to import an additional crate with `StrictMath` (e.g. https://crates.io/crates/rust-strictmath), just for the sake of Gaussian distribution behaving in the same way as Java's implementation does.

If 100% compatibility with the output of the original c-s is required, then I guess that we should.

> 
> NOTICE: this may also be the problem in the future once we decide to seed `pk_seed` distribution with operation's id. This distribution is set via `-pop dist=` parameter, and is used to sample the seeds for `partition_key` generator for each operation.

Reply:
> Isn't `[number of ops]` going to be much higher in practice, i.e. in SCT tests? If so, then the probability rises. If we are going to write with our implementation of c-s and then validate with the original c-s, then this is prone to flakiness.

Yes, that was just an example. The biggest number of ops I've tried so far (locally) is 10M - everything went smoothly.

> If 100% compatibility with the output of the original c-s is required, then I guess that we should.

Maybe @avelanarius will have a stronger opinion on that. I am still not sure about that.